### DISCUSSION TITLE
Ironclad 3 wip

### DISCUSSION THREAD

**Discussion by slozier** [IronLanguages/ironclad#18] [General]
Follow-up of discussion with @BCSharp in https://github.com/IronLanguages/ironclad/issues/16.

@BCSharp , in case you're interested, I pushed some more tweaks to my [wip](https://github.com/slozier/ironclad/tree/wip) branch. I managed to get it running a `tkinter` hello world:
```py
import os
os.environ["TCL_LIBRARY"] = r"C:\Python34\tcl\tcl8.6"

import ironclad
import sys
sys.path.append(r"C:\Python34\DLLs")

import tkinter as tk

root = tk.Tk() 
a = tk.Label(root, text ="Hello World") 
a.pack() 
  
root.mainloop() 
```

**Comment by BCSharp** [comment]
That's great! If you are OK with the commits I have pushed to ipy3-wip, I can rebase your commits on top of it. And if you are satisified with the so-far-code in your wip, then maybe merge it into ipy3-wip?

I have also a wip-on-wip branch named [ipy3-clang](https://github.com/BCSharp/ironclad/tree/ipy3-clang) on which I have experimental SCons code to build `ic_mscvr90.dll` using Clang, with PDB generation. Comments welcome. Once it proves itself (i.e. debugging works as expected), I intend to use Clang for `python34.dll` too.

The biggest challenges with Clang I've encountered:
1. `lld-link` seems to be unable to link directly with a DLL. I used the `gendef`/`dlltool` combo to generate the implib, which works OK, but still does not feel as good as what `ld` is capable of. Or maybe there is an `lld` option for that which I am not aware of.
2. `ld` exports all symbols from a DLL by default. `lld-link` "understands" the switch `-export-all-symbols` (i.e. doesn't bark at it as unrecognized), but still does not want to export all symbols automatically. This is not an issue for `ic_msvcr90.dll`, which is small and is to become even smaller, but for `python34.dll` a DEF will need to be generated during build. Or again, maybe there is a way to get `lld` to behave, but I simply don't know it yet.

**Comment by slozier** [comment]
Alright, I did a bit of cleanup and pushed to ipy3-wip. Unless you have any objections I'm thinking we might as well merge it to the main branch?

As for the ipy3-clang branch, assuming the debugging business all works then great! Better than the `printf` debugging I've been doing. :smile: My knowledge when it comes to C compilers is somewhat lacking so I'm probably not going to be of much help.

**Reply by BCSharp** [reply]
I think it's a good idea to merge it to the main branch. I have just a few questions and one request.

1. The commit getting hashlib (mostly) working introduced a regression: `testAddObjectWithExistingReferenceAddsMappedObjectAndDecRefsPointer` is failing. Are you aware of that?

2. About branch naming: you have mentioned merging it to the main branch. Did you mean the master branch? I have noticed that CPython, NumPy have renamed their 'master' to 'main' and new git repos default now with the name 'main' (which is arguably more readable). If you are considering following this convention, I think now would be a good moment to do the rename.

3. About Ironclad 2.7: although it was never fully working, I think it may still be useful to keep it somehow accessible, if not for reference only. How about renaming branch Docker (which has the last commit before ipy3 commits started, I believe) to '2.7'? I'd still update the Dockerfile on that branch to use the windows image with .NET preinstalled, to make the toolset easily accessible.

4. Going forward, what is your preferred way of working? Working with direct commits onto main or via pull requests? Perhaps something like this: trivial, or boring commits directly to 'main', more interesting changes via a PR, which, after sitting proposed for a day or two, if it doesn't attach any comments, can be merged (or squashed?) to main by the author. I don't want to create unneeded review pressure on either of us.

Request: let the merge to 'main' (or 'master') be a fast-forward, rather than a squash. I'd like to see the history of all the changes, there were plenty of them. Perhaps this is your preference too.

**Reply by slozier** [reply]
1. I don't quite remember, but I think some reference counts were going below zero so I probably disabled the decref. I'll try to take a look at it in the coming days.
2. & 3. Alright, I did the following renaming:
- `ipy3-wip` to `main` and made it the default branch. The `master` branch is now gone.
- `docker` to `2.7`
4. I'm fine with commits to `main` unless you think it's something worth reviewing. Until the repo becomes more stable I'm not super worried about it. I guess I should look into setting up some github actions to run the test suite...

**Reply by slozier** [reply]
1. I reverted the change to get the test passing as I couldn't get it to fail with the code I was testing. But I assume I'll run into it again at some point. 😄

**Comment by BCSharp** [comment]
A suggestion: to change the _About_ text on the GitHub project page to something like:

> CPython API compatibility layer for IronPython 

I got that from the original Google Code page for Ironclad and find more informative than what we have now.

**Reply by slozier** [reply]
Works for me. Done.

**Comment by BCSharp** [comment]
The project root directory contains `license.txt`, which is 14 years old and likely needs to be updated. At least to address the .NET Foundation. The original license seems to be the MIT license, but the standard .NET Foundation license is Apache (that is, as used by IronPython).

Just something to think about.

**Reply by slozier** [reply]
Indeed something to think about. I'm not sure the .NET Foundation aspect covers this project as it was set up specifically with IronPython and the DLR in mind. in any case, I'm fine with sticking to the MIT license for this.

**Reply by GeorgeS2019** [reply]
.NET Foundation prefers MIT license.  If agree, please update so more people could join the contribution

**Comment by slozier** [comment]
@BCSharp Getting an access violation exception since [92006a](https://github.com/IronLanguages/ironclad/commit/92006af7bd12a1761c9cae143de25fcc86d50edd) when trying to load up `_hashlib` , any ideas?

**Reply by BCSharp** [reply]
Well, I was able to reproduce the problem and will look into it tomorrow. I have no convincing ideas as yet. Strange that `test_hashlib` doesn't fail.

**Reply by BCSharp** [reply]
I've fixed the assembly code. `_hashlib` should import now.

**Reply by BCSharp** [reply]
> Strange that `test_hashlib` doesn't fail.

Mystery solved: I was inadvertently using the wrong `test_hashlib` (the one from `Tests`). Will use a different name for the Ironclad test going forward.

**Reply by GeorgeS2019** [reply]
I know those guys at **Ikvm** is also using **clang** to get jdk for.NET6+ working

**Reply by GeorgeS2019** [reply]
@BCSharp 
Is there major block moving Ironclad forwards?

**Reply by BCSharp** [reply]
From my side — availability of time.

**Comment by GeorgeS2019** [comment]
should this project be jumpstarted in 2025?

is there alternative?

**Reply by BCSharp** [reply]
@GeorgeS2019 If you are looking for alternatives to IronPython that will allow running packages that are CPython extensions within .NET programs, then you may want to check out [CSnakes](https://github.com/tonybaloney/csnakes) — and maybe report back your experiences (I have none, but it looks good on paper/YouTube videos 😃 ).

From my side, there is one major roadblock for moving Ironclad forward, and that's that CPython 3.4 depends on MSVCRT 9.0, which is obsolete and unfortunately incompatible with UCRT, which is part of nowadays Windows, and which IronPython uses. CPython 3.6 uses MSVCRT 10.0, which is supposed to be compatible with UCRT. So, practically speaking, we need wo move IronPython towards a 3.6 release first.

**Reply by RyogiMutsuki** [reply]
>we need wo move IronPython towards a 3.6 release first.

👀 

When is IrnoPython scheduled for version 3.6?

I'd like to use IrnoPython on my project, but the current version (3.4.2) doesn't seem to allow me to do so.

**Reply by GeorgeS2019** [reply]
https://github.com/IronLanguages/ironpython3/discussions/1678

**Comment by GeorgeS2019** [comment]
Thx for taking time to share your timeline

**Comment by GeorgeS2019** [comment]
https://github.com/IronLanguages/ironpython3/discussions/1678
